src/cpp/include/openvino/genai/vlm_pipeline.hpp

```diff
@@ -0,0 +1,483 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "openvino/genai/tokenizer.hpp"
+#include "openvino/genai/vlm_minicpmv.hpp"
+#include "openvino/genai/vlm_sampling.hpp"
+#include "openvino/genai/clip.hpp"
+#include <openvino/openvino.hpp>
+#include "../src/text_callback_streamer.hpp"
+#include <optional>
+#include <random>
+
+typedef std::chrono::high_resolution_clock Time;
+typedef std::chrono::nanoseconds ns;
+
+struct Args {
+    bool do_sample = false;
+    int top_k = 0;
+    float top_p = 0.7f;
+    float temp = 0.95f;
+    float repeat_penalty = 1.0f;
+};
+
+int64_t get_out_token_id(const std::vector<int>& input_ids, float* logits, size_t vocab_size, Args args) {
+    int64_t out_token;
+
+    // logits pre-process
+    if (args.repeat_penalty != 1.f) {
+        sampling_repetition_penalty(logits, logits + vocab_size, input_ids, args.repeat_penalty);
+    }
+
+    if (args.do_sample)
+    {
+        if (args.temp > 0) {
+            sampling_temperature(logits, logits + vocab_size, args.temp);
+        }
+
+        std::vector<TokenIdScore> token_scores(vocab_size);
+        for (int i = 0; i < vocab_size; i++) {
+            token_scores[i] = TokenIdScore(i, logits[i]);
+        }
+
+        // top_k sampling
+        if (0 < args.top_k && args.top_k < (int)token_scores.size()) {
+            sampling_top_k(token_scores.data(), token_scores.data() + args.top_k,
+                token_scores.data() + token_scores.size());
+            token_scores.resize(args.top_k);
+        }
+
+        // top_p sampling
+        if (0.f < args.top_p && args.top_p < 1.f) {
+            auto pos = sampling_top_p(token_scores.data(), token_scores.data() + token_scores.size(), args.top_p);
+            token_scores.resize(pos - token_scores.data());
+        }
+
+        // sample next token
+        sampling_softmax_inplace(token_scores.data(), token_scores.data() + token_scores.size());
+        for (size_t i = 0; i < token_scores.size(); i++) {
+            logits[i] = token_scores[i].score;
+        }
+
+        thread_local std::random_device rd;
+        thread_local std::mt19937 gen(rd());
+
+        std::discrete_distribution<> dist(logits, logits + token_scores.size());
+        out_token = token_scores[dist(gen)].id;
+    }
+    else {
+        out_token = std::max_element(logits, logits + vocab_size) - logits;
+    }
+
+    return out_token;
+}
+
+static double get_duration_ms_until_now(Time::time_point& startTime) {
+    return std::chrono::duration_cast<ns>(Time::now() - startTime).count() * 0.000001;
+}
+
+ov::Tensor get_image_embedding(const std::pair<std::vector<std::vector<ov::Tensor>>, size_t>& slices, ov::genai::Tokenizer& tokenizer, ov::InferRequest& embedding, ov::InferRequest& resampler) {
+    auto [image_embed_slices, ratio] = slices;
+    std::string user_prompt;
+    size_t embedding_dim;
+    size_t embedding_len = 0;
+    size_t idx;
+    int scale_emb = 12;
+
+    user_prompt = "<用户>";
```

Comment:
can we define all such magic numbers / strings in capital letters and at the beginning of the file? 
to keep the rest of the code in generic form